actors/builtin/miner/expiration_queue.go

```diff
@@ -0,0 +1,636 @@
+	if err := q.add(newExpiration, noOnTimeSectors, earlySectors, noActivePower, rescheduledPower, noOnTimePledge); err != nil {
+		return NewPowerPairZero(), err
+	}
+
+	return rescheduledPower.Add(expiringPower), nil
+}
+
+// Re-schedules *all* sectors to expire at an early expiration epoch, if they wouldn't expire before then anyway.
+func (q ExpirationQueue) RescheduleAllAsFaults(faultExpiration abi.ChainEpoch) error {
+	var rescheduedEpochs []uint64
+	var rescheduledSectors []*abi.BitField
+	rescheduledPower := NewPowerPairZero()
+
+	var es ExpirationSet
+	if err := q.Array.ForEach(&es, func(e int64) error {
+		epoch := abi.ChainEpoch(e)
+		if epoch <= q.quant.QuantizeUp(faultExpiration) {
+			// Regardless of whether the sectors were expiring on-time or early, all the power is now faulty.
+			// Pledge is still on-time.
+			es.FaultyPower = es.FaultyPower.Add(es.ActivePower)
+			es.ActivePower = NewPowerPairZero()
+			if err := q.mustUpdate(epoch, &es); err != nil {
+				return err
+			}
+		} else {
+			rescheduedEpochs = append(rescheduedEpochs, uint64(epoch))
+			rescheduledSectors = append(rescheduledSectors, es.OnTimeSectors, es.EarlySectors)
+			rescheduledPower = rescheduledPower.Add(es.ActivePower)
+			rescheduledPower = rescheduledPower.Add(es.FaultyPower)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// Add rescheduled sectors to new expiration as early-terminating and faulty.
+	allRescheduled, err := bitfield.MultiMerge(rescheduledSectors...)
+	if err != nil {
+		return xerrors.Errorf("failed to merge rescheduled sectors: %w", err)
+	}
+	noOnTimeSectors := abi.NewBitField()
+	noActivePower := NewPowerPairZero()
+	noOnTimePledge := abi.NewTokenAmount(0)
+	if err = q.add(faultExpiration, noOnTimeSectors, allRescheduled, noActivePower, rescheduledPower, noOnTimePledge); err != nil {
+		return err
+	}
+
+	// Trim the rescheduled epochs from the queue.
+	if err = q.BatchDelete(rescheduedEpochs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Removes sectors from any queue entries in which they appear that are earlier then their scheduled expiration epoch,
+// and schedules them at their expected termination epoch.
+// Pledge for the sectors is re-added as on-time.
+// Power for the sectors is changes from faulty to active (whether rescheduled or not).
+// Returns the newly-recovered power. Fails if any sectors are not found in the queue.
+func (q ExpirationQueue) RescheduleRecovered(sectors []*SectorOnChainInfo, ssize abi.SectorSize) (PowerPair, error) {
+	remaining := make(map[abi.SectorNumber]struct{}, len(sectors))
+	for _, s := range sectors {
+		remaining[s.SectorNumber] = struct{}{}
+	}
+
+	// Traverse the expiration queue once to find each recovering sector and remove it from early/faulty there.
+	var sectorsRescheduled []*SectorOnChainInfo
+	recoveredPower := NewPowerPairZero()
+	if err := q.traverseMutate(func(epoch abi.ChainEpoch, es *ExpirationSet) (changed, keepGoing bool, err error) {
```

Comment:
We should add a comment here explaining that it can process at most 14 epochs (14 proving periods = 14 deadlines epochs = 14 quantized epochs).

It would be nice to assert this as well (even though it should be impossible to exceed that limit). However, _not_ asserting it is actually safer (in case something goes wrong). Maybe log something?